include/inexor/vulkan-renderer/world/collision.hpp

```diff
@@ -0,0 +1,250 @@
+
+#include <array>
+#include <limits>
+#include <memory>
+#include <optional>
+#include <string>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+namespace inexor::vulkan_renderer::world {
+
+/// @brief A wrapper for collisions between a ray and octree geometry.
+/// This class is used for octree collision, but it can be used for every cube-like data structure
+/// @tparam T A templatable type which offers a size() and center() method.
+template <typename T>
+class RayCubeCollision {
+    const T &m_cube;
+
+    using cube_face = std::tuple<std::string, glm::vec3>;
+    using cube_corner = std::tuple<std::string, glm::vec3>;
+    using cube_edge = std::tuple<std::string, glm::vec3>;
+
+    glm::vec3 m_intersection;
+    cube_face m_selected_face;
+    cube_corner m_nearest_corner;
+    cube_edge m_nearest_edge;
+
+public:
+    /// @brief Calculate point of intersection, selected face,
+    /// nearest corner on that face, and nearest edge on that face.
+    /// @param cube The cube to check for collision.
+    /// @param ray_pos The start point of the ray.
+    /// @param ray_dir The direction of the ray.
+    RayCubeCollision(const T &cube, const glm::vec3 ray_pos, const glm::vec3 ray_dir) : m_cube(cube) {
+
+        // This lambda adjusts the center points on a cube's face to the size of the octree,
+        // so collision works with cubes of any size. This does not yet account for rotations!
+        const auto adjust_coordinates = [=](const glm::vec3 pos) {
+            // TODO: Take rotation of the cube into account.
+            return m_cube.center() + pos * (m_cube.size() / 2);
+        };
+
+        const std::array<glm::vec3, 6> cube_directions{
+            glm::vec3(-1.0f, 0.0f, 0.0f), // left
+            glm::vec3(1.0f, 0.0f, 0.0f),  // right
+            glm::vec3(0.0f, -1.0f, 0.0f), // front
+            glm::vec3(0.0f, 1.0f, 0.0f),  // back
+            glm::vec3(0.0f, 0.0f, 1.0f),  // top
+            glm::vec3(0.0f, 0.0f, -1.0f)  // bottom
+        };
+
+        // The coordinates of the center of every face of the cube.
+        const std::array<cube_face, 6> cube_face_centers{cube_face{"left", adjust_coordinates(cube_directions[0])},
+                                                         cube_face{"right", adjust_coordinates(cube_directions[1])},
+                                                         cube_face{"front", adjust_coordinates(cube_directions[2])},
+                                                         cube_face{"back", adjust_coordinates(cube_directions[3])},
+                                                         cube_face{"top", adjust_coordinates(cube_directions[4])},
+                                                         cube_face{"bottom", adjust_coordinates(cube_directions[5])}};
+
+        // The coordinates of every corner of the cube.
+        std::array<cube_corner, 8> cube_corners = {
+            cube_corner{"left back top", adjust_coordinates({-1.0f, 1.0f, 1.0f})},
```

Comment:
Would it be possible to externalize magic string literals?

Reply:
We could put those into a separate data structure, but then we would also need some indexing. I think it's better to keep them linked as `std::tuple`.

These are for debugging purposes only. We must not forget that these might no longer be correct if we take rotations of the octree into account, as "top" might be "bottom" if we rotate 180 degrees around x axis for example. As long as we have no rotations, we should keep them in.

Reply:
This means as soon as we support rotations, we no longer will store these values as strings. We will simply return the index of the face which is being selected and the normal vector on that face.

Reply:
Why not do that now and provide a function which will translate them?

Reply:
I thought about this as well.

Reply:
I will add this naming mechanism when implementing octree rotation.
https://github.com/inexorgame/vulkan-renderer/issues/365